### PR TITLE
Implement schema selection for internet url provider

### PR DESCRIPTION
**Pull request by ProvoK**
### What does this changes

Adds the ability to generate url with control over the scheme.

### What was wrong

It is currently not possible. A random `http` or `https` url is generated.
Related to: https://github.com/joke2k/faker/issues/692

### How this fixes it

Adding optional parameter to internet provider `url` method.

The actual signature change is my proposal, i don't know if it's reasonable or good enough.
I'll do better test cases (and implementation?) if this change in behavior will be accepted.

### CODE REVIEW

**Comment by ProvoK** [comment]
I don't really know why it's not working on certain python versions 😐 . I'll take a look asap.

File: faker/providers/internet/__init__.py [hunk 1]
```diff
@@ -165,8 +165,22 @@ def domain_word(self,):
...
+    def _filter_url_formats(self, exclude_scheme):
+        if not isinstance(exclude_scheme, str):
+            raise ValueError('exclude_scheme param should be a string')
+        if exclude_scheme.lower() not in ['https', 'http']:
+            raise ValueError('exclude_scheme param should either `http` or `https`')
+
+        pattern = '{}://'.format(exclude_scheme)
+        return filter(lambda f: not f.startswith(pattern), self.url_formats)
+
+    def url(self, exclude_scheme=None):
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: I would rather have this more flexible and change the signature to something like:

```python
def url(self, schemes=None):
    if schemes is None:
        schemes = ['http', 'https']
```
This way people can generate any kind of scheme.

Comment (ProvoK) [reply faker/providers/internet/__init__.py]: Sure, I'm totally fine with that :)
I thought `exclude_scheme` were reasonable since actually only `http` and `https` schemes are generated.

I'm going to work on that asap, but I've a couple of questions.
What the expected behavior for the following scenarios?
```python
fake = Faker()

# One or all scheme/s does not "exists". 
fake.url(schemes=['foo', 'http'])   # foo://my-bar.org or error?
fake.url(schemes=['foo', bar'])

# Duplicates, raise an error?
fake.url(schemes=['foo', foo'])
fake.url(schemes=['http', 'http'])
```

Comment (swizzard) [reply faker/providers/internet/__init__.py]: Why does `Faker` need to care about which schemes are real or not?

Comment (fcurella) [reply faker/providers/internet/__init__.py]: @ProvoK:

* `schemes=['foo', bar']` will generate urls with `foo://` or `bar://`. As @swizzard say, `Faker` does't need to care if the scheme is real or not.
* `schemes=['foo', foo']` normalizes to a Set, so somehting like `schemes = Set(schemes)`

Comment (ProvoK) [reply faker/providers/internet/__init__.py]: @swizzard 
Actually, I was questioning about it since i was not sure what would be the best choice. If the generation of the url/email/whatever is too much permissive, a random string could be the same and there would be not the need of faker. However this seems not the case. :)

@fcurella 
got it!

thanks guys :+1:

Comment (ProvoK) [reply faker/providers/internet/__init__.py]: @fcurella 
I've done changes as requested. Tell me if there is something more i should/can do :smile:

File: faker/providers/internet/__init__.py [hunk 2]
```diff
@@ -165,8 +163,12 @@ def domain_word(self,):
     def tld(self):
         return self.random_element(self.tlds)
 
-    def url(self):
-        pattern = self.random_element(self.url_formats)
+    def url(self, schemes=None):
+        if not schemes:
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: use `if schemes is None:`. This way, users can pass an empty list and have schemeless URL ('malformed' urls like that are often used in HTML pages to load a resource regardless of http/https)

File: faker/providers/internet/__init__.py [hunk 3]
```diff
@@ -165,8 +163,12 @@ def domain_word(self,):
...
+            schemes = ['http', 'https']
+
+        pattern = '{}{}'.format(self.random_element(schemes), self.random_element(self.url_formats))
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: make sure `schemes` is not an empty list before calling `self.random_element(schemes)`

File: faker/providers/internet/__init__.py (hunk 2 above)
Comment (ProvoK) [reply faker/providers/internet/__init__.py]: In this case would it be something like `www.domain.com`, and not `://www.domain.com`, i suppose

Comment (fcurella) [reply faker/providers/internet/__init__.py]: No, the expected output for `schemes=[]` would be a protocol-relative URL like `://www.domain.com`

Comment (ProvoK) [reply faker/providers/internet/__init__.py]: Done :+1: I added a bit of documentation too

Comment (ProvoK) [reply faker/providers/internet/__init__.py]: I don't understand how coveralls check had failed, it shows decreased coverage for parts I've not touched or whatever

**Comment by fcurella** [comment]
@ProvoK don't worry about coverage. It's a little falke because the tests are randomized and every once in a while it just so happens that an edge case doesn't get hit.
